File: packages/engine/src/core/peers.ts
import { readdirSync, readFileSync } from 'node:fs'
import { homedir } from 'node:os'
import { join } from 'node:path'

/**
 * Claude Code peer sessions, read from the host's own session registry.
 *
 * sofar is the ADDRESSING layer here and never the transport (D1). Where a
 * collision surface already names the session it collided with, this turns
 * that id into the name Claude Code's own `SendMessage` addresses — and then
 * gets out of the way. sofar never binds the inbox socket, never posts to
 * one, and never calls a model (felt-cost D3): handing an agent an address
 * costs nothing and asks nothing of the network.
 *
 * That division is not just tidiness. Posting into the socket ourselves would
 * mean an undocumented wire format, a verification rule that only exempts a
 * session's own children, inbound controls that hold non-child posts for user
 * approval, and a delivered message billed to the user like a typed prompt.
 * Handing over a name has none of those failure modes, because the agent doing
 * the sending is already inside the host that owns the channel.
 *
 * THE REGISTRY IS UNDOCUMENTED. Claude Code writes one file per live session
 * at `~/.claude/sessions/<pid>.json` carrying — among fields we deliberately
 * ignore — the `sessionId` sofar already stores (the hook JSON's session_id is
 * the same value) and the `name` the host answers to. No published contract
 * promises that shape, so every read here is best-effort in the BD22 sense: a
 * missing directory, an unreadable file, malformed JSON, or a field that
 * changed type all resolve to "no peer", and the caller renders exactly what
 * it rendered before this module existed. Nothing downstream may become
 * correct only when a peer resolves.
 *
 * Everything reads FILES — no subprocess, matching core/git.ts. This sits on
 * the UserPromptSubmit path, inside the 100ms end-to-end shim budget
 * (speed T2), so spawning anything to answer a question a small JSON file
 * already holds would be the most expensive line in the block.
 */

/** One live Claude Code session, as the host's own registry describes it. */
export interface Peer {
  /** Claude Code's session id — the same value sofar stores as a session id. */
  sessionId: string
  /** The name `SendMessage` addresses. */
  name: string
  /** Working directory: the host's own tie-breaker when two names collide. */
  cwd: string
  /**
   * Another live peer answers to the same name. Claude Code derives a default
   * name from the working directory's folder, so several sessions in one repo
   * routinely collide, and the host disambiguates with a short identifier we
   * cannot reconstruct. A caller that renders an ambiguous name must say so
   * rather than imply the bare name will reach anyone in particular.
   */
  ambiguous: boolean
}

/**
 * Most registry files to read in one pass. The directory holds one file per
 * LIVE session, so real-world size is single digits and this never binds; it
 * exists so a host that stopped cleaning up cannot turn a hook line into an
 * unbounded directory walk. Exceeding it degrades the same way every other
 * failure here does — some peers simply do not resolve.
 */
const REGISTRY_SCAN_MAX = 128

function registryDir(env: Record<string, string | undefined>): string {
  const configured = env.CLAUDE_CONFIG_DIR
  if (typeof configured === 'string' && configured.length > 0) {
    return join(configured, 'sessions')
  }
  return join(homedir(), '.claude', 'sessions')
}

interface RegistryEntry {
  sessionId: string
  name: string
  cwd: string
  pid: number
}

/**
 * Parse one registry file, or null. Every field is checked at its use type
 * rather than trusted from the cast: this is a foreign format that can change
 * under us, and a shape drift must read as "no peer", never as a crash on a
 * hook path or an `undefined` rendered into an agent-facing line.
 */
function readEntry(path: string): RegistryEntry | null {
  try {
    const raw: unknown = JSON.parse(readFileSync(path, 'utf8'))
    if (typeof raw !== 'object' || raw === null) return null
    const rec = raw as Record<string, unknown>
    const { sessionId, name, cwd, pid } = rec
    if (typeof sessionId !== 'string' || sessionId.length === 0) return null
    if (typeof name !== 'string' || name.length === 0) return null
    if (typeof cwd !== 'string') return null
    if (typeof pid !== 'number' || !Number.isInteger(pid) || pid <= 0) return null
    return { sessionId, name, cwd, pid }
  } catch {
    return null
  }
}

/**
 * Is the registered process still there? A session killed rather than exited
 * leaves its file behind, and addressing a dead peer spends the agent's turn
 * on a message nobody will read. Signal 0 sends nothing — it only asks whether
 * the pid is addressable — so this stays the no-subprocess check the rest of
 * the module is.
 *
 * PID reuse could in principle answer yes for an unrelated process. It does
 * not matter here: callers only ask about session ids the FOLD already counts
 * as live, so this is a second, independent liveness signal narrowing an
 * already-narrow set, and its worst case is one un-actionable name in a hint.
 */
function alive(pid: number): boolean {
  try {
    process.kill(pid, 0)
    return true
  } catch {
    return false
  }
}

/**
 * Every live peer the registry knows about. Empty whenever the registry is
 * absent, unreadable, or shaped differently than expected — the caller cannot
 * distinguish those cases, and deliberately so: all four mean the same thing
 * at the surface, which is that no address gets rendered.
 */
export function livePeers(env: Record<string, string | undefined> = process.env): Peer[] {
  const dir = registryDir(env)
  let files: string[]
  try {
    files = readdirSync(dir).filter((f) => f.endsWith('.json'))
  } catch {
    return []
  }

  const entries: RegistryEntry[] = []
  for (const file of files.slice(0, REGISTRY_SCAN_MAX)) {
    const entry = readEntry(join(dir, file))
    if (entry !== null && alive(entry.pid)) entries.push(entry)
  }

  // Ambiguity is a property of the whole live set, not of the subset a caller
  // happens to ask about: a name that reaches two sessions is ambiguous even
  // when only one of them collided with us.
  const byName = new Map<string, number>()
  for (const e of entries) byName.set(e.name, (byName.get(e.name) ?? 0) + 1)

  return entries.map((e) => ({
    sessionId: e.sessionId,
    name: e.name,
    cwd: e.cwd,
    ambiguous: (byName.get(e.name) ?? 0) > 1,
  }))
}

/**
 * Peers for the given session ids, keyed by session id. Ids with no live
 * registry entry are simply absent from the map — the caller renders those
 * exactly as it did before, naming the session without an address.
 */
export function resolvePeers(
  sessionIds: readonly string[],
  env: Record<string, string | undefined> = process.env,
): Map<string, Peer> {
  if (sessionIds.length === 0) return new Map()
  const wanted = new Set(sessionIds)
  const found = new Map<string, Peer>()
  for (const peer of livePeers(env)) {
    if (wanted.has(peer.sessionId)) found.set(peer.sessionId, peer)
  }
  return found
}

File: crates/sofar-core/src/peers.rs
//! Live Claude Code peers (`core/peers.ts`, peer-messaging): the session
//! registry under `$CLAUDE_CONFIG_DIR/sessions` (default `~/.claude/sessions`),
//! liveness by pid. `process.kill(pid, 0)` has no std equivalent: Linux asks
//! `/proc/<pid>`, elsewhere ONE `ps -p` per resolution — only ever when a
//! sibling must be named, so a quiet prompt still spawns nothing.

use std::path::PathBuf;

use crate::json::{self, Json};

const REGISTRY_SCAN_MAX: usize = 128;

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Peer {
    pub session_id: String,
    pub name: String,
    pub cwd: String,
    pub ambiguous: bool,
}

struct RegistryEntry {
    session_id: String,
    name: String,
    cwd: String,
    pid: u64,
}

fn registry_dir() -> PathBuf {
    if let Some(configured) = std::env::var_os("CLAUDE_CONFIG_DIR").filter(|v| !v.is_empty()) {
        return PathBuf::from(configured).join("sessions");
    }
    std::env::var_os("HOME")
        .map_or_else(|| PathBuf::from("/"), PathBuf::from)
        .join(".claude")
        .join("sessions")
}

fn read_entry(path: &std::path::Path) -> Option<RegistryEntry> {
    let bytes = std::fs::read(path).ok()?;
    let Json::Obj(rec) = json::parse(&String::from_utf8_lossy(&bytes)).ok()? else {
        return None;
    };
    let session_id = rec.get("sessionId")?.as_nonempty_str()?.to_owned();
    let name = rec.get("name")?.as_nonempty_str()?.to_owned();
    let cwd = rec.get("cwd")?.as_str()?.to_owned();
    let pid = rec.get("pid")?.as_f64()?;
    if !(pid.is_finite() && pid.fract() == 0.0 && pid > 0.0) {
        return None;
    }
    #[allow(
        clippy::cast_possible_truncation,
        clippy::cast_sign_loss,
        reason = "checked positive integer"
    )]
    Some(RegistryEntry {
        session_id,
        name,
        cwd,
        pid: pid as u64,
    })
}

/// Which of these pids are alive.
fn alive_pids(pids: &[u64]) -> Vec<u64> {
    if pids.is_empty() {
        return Vec::new();
    }
    if cfg!(target_os = "linux") {
        return pids
            .iter()
            .copied()
            .filter(|pid| std::path::Path::new(&format!("/proc/{pid}")).exists())
            .collect();
    }
    let list = pids
        .iter()
        .map(u64::to_string)
        .collect::<Vec<_>>()
        .join(",");
    let Ok(out) = std::process::Command::new("ps")
        .args(["-p", &list, "-o", "pid="])
        .stdin(std::process::Stdio::null())
        .stderr(std::process::Stdio::null())
        .output()
    else {
        return Vec::new();
    };
    String::from_utf8_lossy(&out.stdout)
        .split('\n')
        .filter_map(|l| l.trim().parse::<u64>().ok())
        .filter(|p| pids.contains(p))
        .collect()
}

/// `livePeers`: every live registry entry, ambiguity judged over the whole set.
#[must_use]
pub fn live_peers() -> Vec<Peer> {
    let Ok(entries) = std::fs::read_dir(registry_dir()) else {
        return Vec::new();
    };
    let mut files: Vec<PathBuf> = entries
        .filter_map(Result::ok)
        .map(|e| e.path())
        .filter(|p| p.to_string_lossy().ends_with(".json"))
        .collect();
    // `readdirSync` order is the directory's; it only bounds the scan.
    files.truncate(REGISTRY_SCAN_MAX);
    let read: Vec<RegistryEntry> = files.iter().filter_map(|p| read_entry(p)).collect();
    let alive = alive_pids(&read.iter().map(|e| e.pid).collect::<Vec<_>>());
    let live: Vec<&RegistryEntry> = read.iter().filter(|e| alive.contains(&e.pid)).collect();
    live.iter()
        .map(|e| Peer {
            session_id: e.session_id.clone(),
            name: e.name.clone(),
            cwd: e.cwd.clone(),
            ambiguous: live.iter().filter(|o| o.name == e.name).count() > 1,
        })
        .collect()
}

/// `resolvePeers`: the live peers among these session ids.
#[must_use]
pub fn resolve_peers(session_ids: &[String]) -> Vec<Peer> {
    if session_ids.is_empty() {
        return Vec::new();
    }
    live_peers()
        .into_iter()
        .filter(|p| session_ids.contains(&p.session_id))
        .collect()
}
